### ISSUE TITLE
Listing tokens is slow because it fetches all matching tokens

### ISSUE THREAD

**Issue by fredreichbier** [privacyidea/privacyidea#1123]
We sometimes execute redundant SQL queries or queries that are broader than they have to be. If there is a large number of tokens in the system, these redundant SQL queries may have a noticeable impact on performance.

A good way to spot these is to set ``SQLALCHEMY_ECHO = True`` in ``pi.cfg``. All queries are then logged to stderr. In an Apache setup, these are written to the Apache error log.

**Comment by fredreichbier** [comment]
1) Paginated searching for tokens with wildcards inadvertently loads all matching tokens. This line is executed before all requests which carry some ``serial`` parameter (which does not contain `**`)
https://github.com/privacyidea/privacyidea/blob/06b093c911eb7f1b77fb8872c0dc2dc97b2c9ffb/privacyidea/api/before_after.py#L134
and ``get_token_type`` queries all matching tokens:
https://github.com/privacyidea/privacyidea/blob/06b093c911eb7f1b77fb8872c0dc2dc97b2c9ffb/privacyidea/lib/token.py#L456
which can cause problems if we have a quite broad search term. A way to fix this would be to use a ``LIMIT`` clause in ``get_token_type``.

**Comment by fredreichbier** [comment]
On wsgi-based installations, SQLAlchemy pool events and queries can be logged by setting up logging in ``/etc/privacyidea/privacyideaapp.wsgi`` like this:
```python
import sys
sys.stdout = sys.stderr

import logging
handler = logging.FileHandler('/var/log/privacyidea/sqlalchemy.log')
handler.setFormatter(logging.Formatter("[%(asctime)s][%(process)d]"
                                    "[%(thread)d][%(levelname)s]"
                                    "[%(name)s:%(lineno)d] "
                                    "%(message)s"))
logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)
logging.getLogger('sqlalchemy.pool').setLevel(logging.DEBUG)
logging.getLogger('sqlalchemy.engine').addHandler(handler)
logging.getLogger('sqlalchemy.pool').addHandler(handler)

from privacyidea.app import create_app
# Now we can select the config file:
application = create_app(config_name="production", config_file="/etc/privacyidea/pi.cfg")
```

**Comment by cornelinux** [comment]
We can use the configured ``page_size`` as ``LIMIT``. 
(I thought this was once the way like sqlalchemy worked).

**Comment by cornelinux** [comment]
THis is bad.
For code simplicity the ``get_token_type(serial)`` is performed in the ``before`` decorator, so that we do not need to determine the tokentype in each special code of each request.

The "**" in https://github.com/privacyidea/privacyidea/blob/06b093c911eb7f1b77fb8872c0dc2dc97b2c9ffb/privacyidea/api/before_after.py#L133
should avoid, that the ``get_token_type`` is called. As the performance issues occur during searches (which will always look like "*123*"), we might simply change the code to

~~~~python
if serial.startswith("*") and serial.endswith("*"):
    tokentype = None
elif serial:
    tokentype = get_token_type(serial)
else:
    tokentype = None
~~~~

**Comment by cornelinux** [comment]
@fredreichbier the ``get_token_type`` is fixed in #1139.
We currently discuss the sql pooling in another issue.

So what is actually remaining in this issue - or can we close it?

**Comment by fredreichbier** [comment]
Fixed by #1139
